tests/unit/s2n_ktls_test_utils_test.c

```diff
@@ -0,0 +1,136 @@
+    EXPECT_SUCCESS(s2n_blob_init(&test_data_blob, test_data, sizeof(test_data)));
+    EXPECT_OK(s2n_get_public_random_data(&test_data_blob));
+
+    /* Test the mock IO stuffer implementation can send/recv records */
+    {
+        /* Test sending send/recv 0 bytes */
+        {
+            DEFER_CLEANUP(struct s2n_connection *server = s2n_connection_new(S2N_SERVER),
+                    s2n_connection_ptr_free);
+            DEFER_CLEANUP(struct s2n_connection *client = s2n_connection_new(S2N_CLIENT),
+                    s2n_connection_ptr_free);
+            DEFER_CLEANUP(struct s2n_test_ktls_io_pair io_pair = { 0 },
+                    s2n_ktls_io_pair_free);
+            EXPECT_OK(s2n_test_init_ktls_stuffer_io(server, client, &io_pair));
+
+            size_t send_zero = 0;
+            /* Init send msghdr */
+            struct iovec send_msg_iov = { .iov_base = test_data, .iov_len = send_zero };
+            struct msghdr send_msg = { .msg_iov = &send_msg_iov, .msg_iovlen = 1 };
+            /* sendmsg */
+            ssize_t bytes_written = s2n_test_ktls_sendmsg_stuffer_io(server, &send_msg, S2N_TEST_RECORD_TYPE);
+            EXPECT_EQUAL(bytes_written, send_zero);
+
+            EXPECT_EQUAL(io_pair.client_in.send_recv_msg_invoked_count, 1);
+            EXPECT_EQUAL(io_pair.server_in.send_recv_msg_invoked_count, 0);
+        };
+
+        /* Test send/recv non-zero values. Zero is a special case and tested above */
+        for (size_t to_send = 1; to_send < S2N_TLS_MAXIMUM_FRAGMENT_LENGTH; to_send++) {
+            DEFER_CLEANUP(struct s2n_connection *server = s2n_connection_new(S2N_SERVER),
+                    s2n_connection_ptr_free);
+            DEFER_CLEANUP(struct s2n_connection *client = s2n_connection_new(S2N_CLIENT),
+                    s2n_connection_ptr_free);
+            DEFER_CLEANUP(struct s2n_test_ktls_io_pair io_pair = { 0 },
+                    s2n_ktls_io_pair_free);
+            EXPECT_OK(s2n_test_init_ktls_stuffer_io(server, client, &io_pair));
+
+            /* Init send msghdr */
+            struct iovec send_msg_iov = { .iov_base = test_data, .iov_len = to_send };
+            struct msghdr send_msg = { .msg_iov = &send_msg_iov, .msg_iovlen = 1 };
+            /* sendmsg */
+            ssize_t bytes_written = s2n_test_ktls_sendmsg_stuffer_io(server, &send_msg, S2N_TEST_RECORD_TYPE);
+            EXPECT_EQUAL(bytes_written, to_send);
+
+            /* confirm sent data */
+            EXPECT_OK(s2n_test_validate_data(&io_pair.client_in, test_data, to_send));
+            /* confirm sent ancillary data */
+            EXPECT_OK(s2n_test_validate_ancillary(&io_pair.client_in, S2N_TEST_RECORD_TYPE, to_send));
+            /* rewind the read so that the recvmsg can retrieve it */
+            EXPECT_SUCCESS(s2n_stuffer_rewind_read(&io_pair.client_in.data_buffer, to_send));
+            EXPECT_SUCCESS(s2n_stuffer_rewind_read(&io_pair.client_in.ancillary_buffer, S2N_TEST_KTLS_MOCK_HEADER_SIZE));
```

Comment:
Shouldn't the rewind be handled by s2n_test_validate_ancillary? Alternatively, s2n_test_validate_ancillary could work on a copy of the stuffer to avoid messing with its read/write pointers.

Same with s2n_test_validate_data. Probably shouldn't modify anything.

Reply:
Ill experiment with different things here. Its possibly nicer to read and validate instead of validate in-place. We would just need to update the fn signature to make it more obvious